### PR TITLE
atomic registration

### PR DESCRIPTION
**Pull request by michaelkirk**
Three related changes here - each with their own commit.

PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/ProfileFetcherJob.swift [hunk 1]
```diff
@@ -12,34 +13,54 @@ class ProfileFetcherJob: NSObject {
...
     // This property is only accessed on the main queue.
     static var fetchDateMap = [String: Date]()
 
     public class func run(thread: TSThread, networkManager: TSNetworkManager) {
-        ProfileFetcherJob(thread: thread, networkManager: networkManager).run()
+        ProfileFetcherJob(networkManager: networkManager).run(thread: thread)
     }
 
-    init(thread: TSThread, networkManager: TSNetworkManager) {
+    init(networkManager: TSNetworkManager) {
```
Comment (michaelkirk) [inline Signal/src/ProfileFetcherJob.swift]: Reused the ProfileFetcherJob, but now it doesn't *always* take a thread.

File: SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m [hunk 2]
```diff
@@ -5,17 +5,27 @@
...
++ (nullable PreKeyBundle *)preKeyBundleFromDictionary:(NSDictionary *)dictionary forDeviceNumber:(NSNumber *)number
+{
     PreKeyBundle *bundle        = nil;
-    NSString *identityKeyString = [dictionary objectForKey:@"identityKey"];
-    NSArray *devicesArray       = [dictionary objectForKey:@"devices"];
 
-    if (!(identityKeyString && [devicesArray isKindOfClass:[NSArray class]])) {
-        DDLogError(@"Failed to get identity key or messages array from server request");
+    id identityKeyObject = [dictionary objectForKey:@"identityKey"];
+    if (![identityKeyObject isKindOfClass:[NSString class]]) {
```
Comment (michaelkirk) [inline SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m]: prevent crash when messaging someone with an incomplete profile

File: SignalServiceKit/src/Account/TSAccountManager.m [hunk 3]
```diff
@@ -234,6 +234,11 @@ - (void)verifyAccountWithCode:(NSString *)verificationCode
                                                                                forNumber:phoneNumber
                                                                             signalingKey:signalingKey
                                                                                  authKey:authToken];
+    void (^completedRegistrationBlock)() = ^{
```
Comment (michaelkirk) [inline SignalServiceKit/src/Account/TSAccountManager.m]: Don't consider registration complete until prekeys are uploaded.

File: Signal/src/environment/OWS106EnsureProfileComplete.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import PromiseKit
+
+class OWS106EnsureProfileComplete: OWSDatabaseMigration {
```
Comment (michaelkirk) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Fixup any half-registered users in the wild.

File: Signal/src/environment/OWS106EnsureProfileComplete.swift [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+                var isCompleted = false
+                strongSelf.ensureProfileComplete().then { _ -> Void in
+                    guard isCompleted == false else {
+                        Logger.info("Already saved. Skipping redundant call.")
+                        return
+                    }
+                    Logger.info("\(strongSelf.TAG) complete. Canceling timer and saving.")
+                    isCompleted = true
+                    timer.invalidate()
```
Comment (charlesmchen) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Will this work?  I don't think the local `timer` variable is initialized at the time its captured, so I'd expect this not to work.  Swift might be different in this regard that Obj-C.

File: Signal/src/environment/OWS106EnsureProfileComplete.swift [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                    strongSelf.completionHandler()
+                }.catch { error in
+                    Logger.error("\(strongSelf.TAG) failed with \(error). We'll try again in \(CompleteRegistrationFixerJob.kRetryInterval) seconds.")
+                }.retainUntilComplete()
+            }
+            self.timer = timer
+
+            timer.fire()
```
Comment (charlesmchen) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Why is this necessary?

File: Signal/src/environment/OWS106EnsureProfileComplete.swift [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    // Overriding runUp since we have some specific completion criteria which
+    // is more likely to fail since it involves network requests.
+    override func runUp() {
+        CompleteRegistrationFixerJob(completionHandler: {
+            Logger.info("\(self.TAG) Completed. Saving.")
+            self.save()
+        }).start()
+    }
```
Comment (charlesmchen) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Won't CompleteRegistrationFixerJob get deallocated as soon as it starts?  Who maintains a strong reference to it while it "runs"?

File: SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m (hunk 2 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m]: Are we making the same fix in android and desktop?

File: Signal/src/environment/OWS106EnsureProfileComplete.swift (hunk 6 above)
Comment (michaelkirk) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Not strictly necessary, just makes the initial run *now* instead of in **n** seconds.

File: SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m (hunk 2 above)
Comment (michaelkirk) [inline SignalServiceKit/src/Messages/PreKeyBundle+jsonDict.m]: In my testing Android and Desktop didn't crash in this scenario, so presumably they already have more robust parsing logic than we did.

File: Signal/src/environment/OWS106EnsureProfileComplete.swift (hunk 5 above)
Comment (michaelkirk) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: sorry this was convoluted. This timer is not `self.timer` it's the block param of the same name. I'm going to clean this up a bit to be a little clearer.

Comment (charlesmchen) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: Even still, there may be an issue here.

Comment (charlesmchen) [inline Signal/src/environment/OWS106EnsureProfileComplete.swift]: LGTM with these changes.

**Comment by michaelkirk** [comment]
in bf82d2e2e03ecd1747bbee60592c98fd9f59bc68
